src/System.Collections.Specialized/src/System/Collections/Specialized/OrderedDictionary.cs

```diff
@@ -175,21 +172,27 @@ object ICollection.SyncRoot
         {
             get
             {
-                return ((DictionaryEntry)objectsArray[index]).Value;
+                if (_objectsArray == null)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(index));
+                }
```

Comment:
Why is this beneficial?  Seems like it should just be:
```C#
EnsureObjectsArray();
return ((DictionaryEntry)_objectsArray[index]).Value;
```
And let the ArrayList's indexer throw the appropriate exception.

Reply:
This problem is described in #31972. In PowerShell Core repository, we found that we make tons of unnecessary allocations. Perhaps there are _other applications_ that are prone to this problem. This fix does not require resources. So fix all the properties of the class makes sense.

Reply:
I understand the general concern. I was asking about this case specifically. Going out of our way, including trying to mimic error handling of ArrayList, in order to prevent an allocation when we're about to throw an exception (which is way more expensive) seems to me like overkill.

Reply:
If I remember correctly, it was just the indexing that caused the problem in PowerShell Core.

Reply:
That doesn't make sense to me.  For both the getter and the setter, either the operation is successful and requires that the objects be allocated (as they were previous to this change and still are after this change), or the operation fails, allocates an exception, and throws that exception.  I expect the extra ArrayList allocation in such a failure case is a drop in the bucket when compared to the exception allocation and throw, and even if it weren't, the right fix for PowerShell would be to avoid incorrect usage of the collection in the first place, so that it could avoid the exception.

So, from my perspective, adding this code is trying to microoptimize an already super expensive failure path, and in exchange is increasing the size of the binary and duplicating code from ArrayList that now needs to be maintained.

I'm not objecting to the overall PR, just to these specific pieces of it.

cc: @danmosemsft

Reply:
> If team wants to keep the indexer super fast then I'm not against this because now we can use other properties to determine that the object is empty

I think we're still talking past each other.

Both before and after your change, accessing the this[int] getter and setter with a valid index and the objects already initialized will not incur additional allocations.

Both before and after your change, accessing the this[int] getter with a valid index won't incur additional allocations, because by definition the index is only valid if the objects have been initialized.

Both before and after your change, accessing the this[int] setter with a valid index and the objects not already initialized will force them to be initialized, incurring allocations.

Both before and after your change, accessing the this[int] getter and setter with an invalid index and the objects already initialized will not incur additional allocations.

The only allocation difference for the this[int] indexer from your change, then, is when an invalid index is used and the objects haven't been initialized, in which case your change prevents the allocation... but in that case, we're allocating and throwing an exception, which is way more expensive.  I'm simply saying we shouldn't be spending code to optimize that failure case.

Reply:
Sorry, I understand that I don't understand something :-)

> Even after your change, there is still an allocation on this[int] if the object isn't yet initialized, just as there was before.

Before we allocate ArrayList and Exception. After we allocate only Exception.

Reply:
> Before we allocate ArrayList and Exception. After we allocate only Exception.

That's true, but it also throws the exception.  The cost of allocating and throwing the exception is orders of magnitude more expensive than allocating the ArrayList.  Not only is the ArrayList 40 bytes and ArgumentOutOfRangeException 232 bytes, but `try { throw new ArgumentOutOfRangeException("index"); } catch {}` is, on my machine, ~3,833x more expensive than the ArrayList allocation in terms of throughput.  The ArrayList allocation is completely inconsequential.  Not only that, but this kind of optimization is really only important when it's on a hot path, and if you're throwing exceptions on such a hot path, you have way, way, way bigger problems; the extra ArrayList allocation doesn't matter in such a case. Now, any improvement, regardless of size, can be worthwhile if there's no negatives associated with it.  But that's also rarely the case.  Here, more code is needed, and thus needs to be maintained.

Reply:
> Could you please clarify how do you measure the size of objects?

A variety of ways:
- You can look at the object layout in a debugger like windbg + SOS.
- You can use a tool like sharplab that does that for you, e.g. https://sharplab.io/#v2:C4LgTgrgdgPgAgBgARwIwG4CwAoOAmBHOAZhVQDYU8yB2JAbxyWZVLUrgBYkBZAQwCWUABQBKBkxZSAklADOABwCmAY2AA6ABJK+C4VCUB3JAEEwAcwgBbJVGAB5CA4BmAJT5RzSgKIAPFUoKwAIA9iIARLIAJkq+4aKiWNhSAL44KUA
- You can use a tool like Benchmark.NET and its memory diagnoser, e.g. https://adamsitnik.com/the-new-Memory-Diagnoser/
- You can write code that uses https://apisof.net/catalog/System.GC.GetAllocatedBytesForCurrentThread(), checking the value, allocating the object (ideally a bunch of times and then later dividing), and checking the value again.

- You can do a back-of-the-hand calculation based on looking at the object's fields.

Etc.